### PR TITLE
Fire UIA Events for Output and Cursor Movement

### LINKED ISSUE #2447: Accessibility: Signaling Model
# Description of the new feature/enhancement

UIA has signal events ([link](https://docs.microsoft.com/en-us/windows/win32/winauto/uiauto-event-ids)). ConHost supported them. WindowUiaProvider was an integral part to that.

Windows Terminal currently doesn't support them. We have some architecture in place to do so though.

# Proposed technical implementation details (optional)

WindowUiaProvider already has the groundwork for that. It might be that we don't even need a WindowUiaProvider and just have XAML do some of it.

For reference, the signaling model for conhost was as follows:

<ConHost Event (i.e.: new text)>
--> Signal WindowUiaProvider
--> Signal ScreenInfoUiaProvider
--> Raise UiaAutomationEvent to Client

The client then responds to these events with more actions (i.e.: recreate object, etc...)

EDIT: The functionality for signaling includes...
- [x] Selection Events (selection changed) [PR](https://github.com/microsoft/terminal/pull/2989)
- [ ] Text Buffer Events (new output)
- [ ] Scroll Events

I may possible need...
- [ ] Resize events
- [ ] Cursor Events

### LINKED ISSUE #3791: Text Cursor Indicator does not move with cursor
# Environment

```none
Windows build number: 10.019033.0
Windows Terminal version (if applicable): 0.7.3291.0

Any other software? -
```

# Steps to reproduce

Turn text cursor indicator in windows settings, and start wt

# Expected behavior

Text Cursor Indicator follows text cursor, works fine in Windows Console

# Actual behavior

Text cursor is displayed, but does not follow cursor
Annotation 2019-12-01 085324

### PR DESCRIPTION
**Pull request by carlos-zamora**
## Summary of the Pull Request
This notifies automation clients (i.e.: NVDA, narrator, etc...) of new output being rendered to the screen.

## References
Close #2447 - Signaling for new output and cursor
Close #3791 - fixed by signaling cursor changes

## Detailed Description of the Pull Request / Additional comments
- Added tracing for UiaRenderer. This makes it easier to debug issues with notifying an automation client.
- Fire TextChanged automation events when new content is output to the screen.

## Validation Steps Performed
Verified with NVDA [1]

## Narrator
Narrator works _better_, but is unable to detect new output consistently. There is no harm for narrator when this change goes in.

[1] https://github.com/microsoft/terminal/issues/2447#issuecomment-595879890

### CODE REVIEW

**Comment by carlos-zamora** [comment]
I meant to publish this PR as a draft but I guess there's no undo button...

**Comment by codeofdusk** [comment]
This PR allows NVDA to fully support Terminal. Corresponding NVDA-side changes are in nvaccess/nvda#10784.

File: src/types/UiaTracing.cpp [hunk 1]
```diff
@@ -42,7 +42,7 @@ try
     stream << " _end: { " << end.X << ", " << end.Y << " }";
     stream << " _degenerate: " << utr.IsDegenerate();
     stream << " _wordDelimiters: " << utr._wordDelimiters;
-    stream << " content: " << utr._getTextValue();
+    //stream << " content: " << utr._getTextValue();
```
Comment (zadjii-msft) [inline src/types/UiaTracing.cpp]: Should this be deleted?

Comment (carlos-zamora) [reply src/types/UiaTracing.cpp]: My bad. This was from before Tracing became optional. Fixed.

**Comment by carlos-zamora** [comment]
Given the latest commit, looks like NVDA 2020.1 works fine.
But Narrator gets worse and crashes on any input...

**Comment by miniksa** [comment]
This is me noting that I saw this PR and am not reviewing at this time because @carlos-zamora says there's still some nasty crash.

**Comment by codeofdusk** [comment]
Any chance this PR could be merged without 522b99a7 and the rest of the event support (for Narrator) Could be added in subsequent PRs? This would allow NVDA to support Terminal.

File: src/cascadia/TerminalControl/TermControlAutomationPeer.cpp [hunk 2]
```diff
@@ -72,9 +75,10 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
     // - <none>
     void TermControlAutomationPeer::SignalCursorChanged()
     {
+        UiaTracing::Signal::CursorChanged();
         Dispatcher().RunAsync(Windows::UI::Core::CoreDispatcherPriority::Normal, [&]() {
             // The event that is raised when the text was changed in an edit control.
-            RaiseAutomationEvent(AutomationEvents::TextEditTextChanged);
+            RaiseAutomationEvent(AutomationEvents::TextPatternOnTextSelectionChanged);
```
Comment (leonMSFT) [inline src/cascadia/TerminalControl/TermControlAutomationPeer.cpp]: Just curious, what does this event mean, and why was `TextEditTextChanged` the wrong one?

Comment (carlos-zamora) [reply src/cascadia/TerminalControl/TermControlAutomationPeer.cpp]: TextEditChanged is apparently for composition. Though you _can_ send a TextEditTextChanged event on it's on, the first question is "was it compositionCompleted, started, or what?". For whatever known reason, if that info isn't provided, some other apps still try to read it and get a nullptr exception.

NVDA does not make this mistake so props to them haha.

Comment (zadjii-msft) [reply src/cascadia/TerminalControl/TermControlAutomationPeer.cpp]: I can't +1 this enough

**Comment by codeofdusk** [comment]
Any idea when this PR will make it to Windows Store users?

**Comment by DHowett-MSFT** [comment]
Aiming for tomorrow.

**Comment by codeofdusk** [comment]
Wow, that was good timing!

**Pull request merged**